File: src/CandidatosCore.Core/Dominio/Candidato.cs
namespace CandidatosCore.Core.Dominio;

using CandidatosCore.Core.Fases;

public sealed class Candidato
{
    public string Nombre { get; }
    public string Correo { get; }

    private FaseBase faseActual;

    public Candidato(string nombre, string correo)
    {
        Nombre = nombre;
        Correo = correo;
        faseActual = new Aplicado();
        faseActual.SetContext(this);
    }

    public void Avanzar() => faseActual.Avanzar();

    public void Rechazar() => faseActual.Rechazar();

    public void CambiarFase(FaseBase nuevaFase)
    {
        nuevaFase.SetContext(this);
        faseActual = nuevaFase;
    }

    public FaseBase ObtenerFaseActual() => faseActual;
}

File: src-ts/CandidatosCore.Core/Dominio/Candidato.ts
import { FaseBase, Aplicado } from "../Fases";

export class Candidato {
    public readonly nombre: string;
    public readonly correo: string;
    private faseActual!: FaseBase;

    constructor(nombre: string, correo: string) {
        this.nombre = nombre;
        this.correo = correo;
        this.cambiarFase(new Aplicado());
    }

    public avanzar(): void {
        this.faseActual.avanzar();
    }

    public rechazar(): void {
        this.faseActual.rechazar();
    }

    public cambiarFase(nuevaFase: FaseBase): void {
        nuevaFase.setContext(this);
        this.faseActual = nuevaFase;
    }

    public obtenerFaseActual(): FaseBase {
        return this.faseActual;
    }
}
